roachtest: sqlsmith invalid interface conversion from DBytes to DGeometry

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2785193&tab=buildLog) on [release-21.1@ba54410af79c0481e1e8f8fc45a9b132ee09a2b4](https://github.com/cockroachdb/cockroach/commits/ba54410af79c0481e1e8f8fc45a9b132ee09a2b4):

```
The test failed on branch=release-21.1, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: interface conversion: tree.Datum is *tree.DBytes, not *tree.DGeometry
		stmt:
		WITH
			with_238 (col_1444) AS (SELECT * FROM (VALUES ('f':::rand_typ_2)) AS tab_631 (col_1444)),
			with_239 (col_1445, col_1446)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									'-13 years -8 mons -572 days -12:56:55.072675':::INTERVAL,
									COALESCE(
										((('1987-10-03':::DATE::DATE || NULL::DATE[])::DATE[]::DATE[] || ARRAY['1980-03-13':::DATE,'1989-01-13':::DATE,'1982-02-27':::DATE]::DATE[])::DATE[]::DATE[] || ARRAY['1970-08-22':::DATE,'1974-01-12':::DATE]::DATE[])::DATE[],
										ARRAY['1981-10-15':::DATE,'1985-10-11':::DATE]
									)
								)
						)
							AS tab_632 (col_1445, col_1446)
				)
		SELECT
			count(*) AS col_1447
		FROM
			defaultdb.public.table1@[0] AS tab_633
		WHERE
			true
		GROUP BY
			tab_633.col1_0
		HAVING
			st_containsproperly(tab_633.col1_0::GEOMETRY, '0103000000010000000400000080243A9AF5B79F41144F37B249FBF541F09535D6F9A9CB414C8DB4EEA99BF541C678E01C19C1F1C174584659C4D1FD4180243A9AF5B79F41144F37B249FBF541':::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			53:::INT8;

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2785193&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

Failure is:

```
06:12:04 test.go:344: test failure: 	sqlsmith.go:198,sqlsmith.go:228,test_runner.go:767: error: pq: internal error: interface conversion: tree.Datum is *tree.DBytes, not *tree.DGeometry
		stmt:
		WITH
			with_238 (col_1444) AS (SELECT * FROM (VALUES ('f':::rand_typ_2)) AS tab_631 (col_1444)),
			with_239 (col_1445, col_1446)
				AS (
					SELECT
						*
					FROM
						(
							VALUES
								(
									'-13 years -8 mons -572 days -12:56:55.072675':::INTERVAL,
									COALESCE(
										((('1987-10-03':::DATE::DATE || NULL::DATE[])::DATE[]::DATE[] || ARRAY['1980-03-13':::DATE,'1989-01-13':::DATE,'1982-02-27':::DATE]::DATE[])::DATE[]::DATE[] || ARRAY['1970-08-22':::DATE,'1974-01-12':::DATE]::DATE[])::DATE[],
										ARRAY['1981-10-15':::DATE,'1985-10-11':::DATE]
									)
								)
						)
							AS tab_632 (col_1445, col_1446)
				)
		SELECT
			count(*) AS col_1447
		FROM
			defaultdb.public.table1@[0] AS tab_633
		WHERE
			true
		GROUP BY
			tab_633.col1_0
		HAVING
			st_containsproperly(tab_633.col1_0::GEOMETRY, '0103000000010000000400000080243A9AF5B79F41144F37B249FBF541F09535D6F9A9CB414C8DB4EEA99BF541C678E01C19C1F1C174584659C4D1FD4180243A9AF5B79F41144F37B249FBF541':::GEOMETRY::GEOMETRY)::BOOL
		LIMIT
			53:::INT8;
```

--- a/pkg/sql/mutations/BUILD.bazel
+++ b/pkg/sql/mutations/BUILD.bazel
@@ -9,7 +9,9 @@ go_library(
     importpath = "github.com/cockroachdb/cockroach/pkg/sql/mutations",
     visibility = ["//visibility:public"],
     deps = [
+        "//pkg/geo/geoindex",
         "//pkg/sql/catalog/colinfo",
+        "//pkg/sql/catalog/descpb",
         "//pkg/sql/parser",
         "//pkg/sql/rowenc",
         "//pkg/sql/sem/tree",

--- a/pkg/sql/mutations/mutations.go
+++ b/pkg/sql/mutations/mutations.go
@@ -18,7 +18,9 @@ import (
 	"sort"
 	"strings"
 
+	"github.com/cockroachdb/cockroach/pkg/geo/geoindex"
 	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/descpb"
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/sql/rowenc"
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
@@ -207,6 +209,10 @@ func statisticsMutator(
 			if col == nil {
 				return
 			}
+			// Do not create a histogram 20% of the time.
+			if rng.Intn(5) == 0 {
+				return
+			}
 			colType := tree.MustBeStaticallyKnownType(col.Type)
 			h := randHistogram(rng, colType)
 			stat := colStats[col.Name]
@@ -273,26 +279,31 @@ func statisticsMutator(
 }
 
 // randHistogram generates a histogram for the given type with random histogram
-// buckets.
+// buckets. If colType is inverted indexable then the histogram bucket upper
+// bounds are byte-encoded inverted index keys.
 func randHistogram(rng *rand.Rand, colType *types.T) stats.HistogramData {
-	h := stats.HistogramData{
-		ColumnType: colType,
+	histogramColType := colType
+	if colinfo.ColumnTypeIsInvertedIndexable(colType) {
+		histogramColType = types.Bytes
 	}
-
-	// TODO(mgartner): Generate histogram buckets for JSON columns.
-	if colType.Family() == types.JsonFamily {
-		return h
+	h := stats.HistogramData{
+		ColumnType: histogramColType,
 	}
 
 	// Generate random values for histogram bucket upper bounds.
 	var encodedUpperBounds [][]byte
 	for i, numDatums := 0, rng.Intn(10); i < numDatums; i++ {
 		upper := rowenc.RandDatum(rng, colType, false /* nullOk */)
-		enc, err := rowenc.EncodeTableKey(nil, upper, encoding.Ascending)
-		if err != nil {
-			panic(err)
+		if colinfo.ColumnTypeIsInvertedIndexable(colType) {
+			encs := encodeInvertedIndexHistogramUpperBounds(colType, upper)
+			encodedUpperBounds = append(encodedUpperBounds, encs...)
+		} else {
+			enc, err := rowenc.EncodeTableKey(nil, upper, encoding.Ascending)
+			if err != nil {
+				panic(err)
+			}
+			encodedUpperBounds = append(encodedUpperBounds, enc)
 		}
-		encodedUpperBounds = append(encodedUpperBounds, enc)
 	}
 
 	// Return early if there are no upper-bounds.
@@ -334,6 +345,43 @@ func randHistogram(rng *rand.Rand, colType *types.T) stats.HistogramData {
 	}
 
 	return h
+}
+
+// encodeInvertedIndexHistogramUpperBounds returns a slice of byte-encoded
+// inverted index keys that are created from val.
+func encodeInvertedIndexHistogramUpperBounds(colType *types.T, val tree.Datum) (encs [][]byte) {
+	var keys [][]byte
+	var err error
+	switch colType.Family() {
+	case types.GeometryFamily:
+		tempIdx := descpb.IndexDescriptor{
+			GeoConfig: *geoindex.DefaultGeometryIndexConfig(),
+		}
+		keys, err = rowenc.EncodeGeoInvertedIndexTableKeys(val, nil, &tempIdx)
+	case types.GeographyFamily:
+		tempIdx := descpb.IndexDescriptor{
+			GeoConfig: *geoindex.DefaultGeographyIndexConfig(),
+		}
+		keys, err = rowenc.EncodeGeoInvertedIndexTableKeys(val, nil, &tempIdx)
+	default:
+		keys, err = rowenc.EncodeInvertedIndexTableKeys(val, nil, descpb.EmptyArraysInInvertedIndexesVersion)
+	}
+
+	if err != nil {
+		panic(err)
+	}
+
+	var da rowenc.DatumAlloc
+	for i := range keys {
+		// Each key much be a byte-encoded datum so that it can be
+		// decoded in JSONStatistic.SetHistogram.
+		enc, err := rowenc.EncodeTableKey(nil, da.NewDBytes(tree.DBytes(keys[i])), encoding.Ascending)
+		if err != nil {
+			panic(err)
+		}
+		encs = append(encs, enc)
+	}
+	return encs
 }
 
 // randNumRangeAndDistinctRange returns two random numbers to be used for

--- a/pkg/sql/mutations/mutations.go
+++ b/pkg/sql/mutations/mutations.go
@@ -207,58 +207,8 @@ func statisticsMutator(
 			if col == nil {
 				return
 			}
-			n := rng.Intn(10)
-			seen := map[string]bool{}
 			colType := tree.MustBeStaticallyKnownType(col.Type)
-			// The JSON family does not have a key encoding.
-			if colType.Family() == types.JsonFamily {
-				return
-			}
-			h := stats.HistogramData{
-				ColumnType: colType,
-			}
-			for i := 0; i < n; i++ {
-				upper := rowenc.RandDatumWithNullChance(rng, colType, 0)
-				if upper == tree.DNull {
-					continue
-				}
-				enc, err := rowenc.EncodeTableKey(nil, upper, encoding.Ascending)
-				if err != nil {
-					panic(err)
-				}
-				if es := string(enc); seen[es] {
-					continue
-				} else {
-					seen[es] = true
-				}
-				numRange := randNonNegInt(rng)
-				var distinctRange float64
-				// distinctRange should be <= numRange.
-				switch rng.Intn(3) {
-				case 0:
-					// 0
-				case 1:
-					distinctRange = float64(numRange)
-				default:
-					distinctRange = rng.Float64() * float64(numRange)
-				}
-
-				h.Buckets = append(h.Buckets, stats.HistogramData_Bucket{
-					NumEq:         randNonNegInt(rng),
-					NumRange:      numRange,
-					DistinctRange: distinctRange,
-					UpperBound:    enc,
-				})
-			}
-			sort.Slice(h.Buckets, func(i, j int) bool {
-				return bytes.Compare(h.Buckets[i].UpperBound, h.Buckets[j].UpperBound) < 0
-			})
-			// The first bucket must have numrange = 0, and thus
-			// distinctrange = 0 as well.
-			if len(h.Buckets) > 0 {
-				h.Buckets[0].NumRange = 0
-				h.Buckets[0].DistinctRange = 0
-			}
+			h := randHistogram(rng, colType)
 			stat := colStats[col.Name]
 			if err := stat.SetHistogram(&h); err != nil {
 				panic(err)
@@ -287,9 +237,13 @@ func statisticsMutator(
 					makeHistogram(def)
 				}
 			case *tree.IndexTableDef:
+				// TODO(mgartner): We should make a histogram for each indexed
+				// column.
 				makeHistogram(cols[def.Columns[0].Column])
 			case *tree.UniqueConstraintTableDef:
 				if !def.WithoutIndex {
+					// TODO(mgartner): We should make a histogram for each
+					// column in the unique constraint.
 					makeHistogram(cols[def.Columns[0].Column])
 				}
 			}
@@ -316,6 +270,86 @@ func statisticsMutator(
 		}
 	}
 	return stmts, changed
+}
+
+// randHistogram generates a histogram for the given type with random histogram
+// buckets.
+func randHistogram(rng *rand.Rand, colType *types.T) stats.HistogramData {
+	h := stats.HistogramData{
+		ColumnType: colType,
+	}
+
+	// TODO(mgartner): Generate histogram buckets for JSON columns.
+	if colType.Family() == types.JsonFamily {
+		return h
+	}
+
+	// Generate random values for histogram bucket upper bounds.
+	var encodedUpperBounds [][]byte
+	for i, numDatums := 0, rng.Intn(10); i < numDatums; i++ {
+		upper := rowenc.RandDatum(rng, colType, false /* nullOk */)
+		enc, err := rowenc.EncodeTableKey(nil, upper, encoding.Ascending)
+		if err != nil {
+			panic(err)
+		}
+		encodedUpperBounds = append(encodedUpperBounds, enc)
+	}
+
+	// Return early if there are no upper-bounds.
+	if len(encodedUpperBounds) == 0 {
+		return h
+	}
+
+	// Sort the encoded upper-bounds.
+	sort.Slice(encodedUpperBounds, func(i, j int) bool {
+		return bytes.Compare(encodedUpperBounds[i], encodedUpperBounds[j]) < 0
+	})
+
+	// Remove duplicates.
+	dedupIdx := 1
+	for i := 1; i < len(encodedUpperBounds); i++ {
+		if !bytes.Equal(encodedUpperBounds[i], encodedUpperBounds[i-1]) {
+			encodedUpperBounds[dedupIdx] = encodedUpperBounds[i]
+			dedupIdx++
+		}
+	}
+	encodedUpperBounds = encodedUpperBounds[:dedupIdx]
+
+	// Create a histogram bucket for each encoded upper-bound.
+	for i := range encodedUpperBounds {
+		// The first bucket must have NumRange = 0, and thus DistinctRange = 0
+		// as well.
+		var numRange int64
+		var distinctRange float64
+		if i > 0 {
+			numRange, distinctRange = randNumRangeAndDistinctRange(rng)
+		}
+
+		h.Buckets = append(h.Buckets, stats.HistogramData_Bucket{
+			NumEq:         randNonNegInt(rng),
+			NumRange:      numRange,
+			DistinctRange: distinctRange,
+			UpperBound:    encodedUpperBounds[i],
+		})
+	}
+
+	return h
+}
+
+// randNumRangeAndDistinctRange returns two random numbers to be used for
+// NumRange and DistinctRange fields of a histogram bucket.
+func randNumRangeAndDistinctRange(rng *rand.Rand) (numRange int64, distinctRange float64) {
+	numRange = randNonNegInt(rng)
+	// distinctRange should be <= numRange.
+	switch rng.Intn(3) {
+	case 0:
+		distinctRange = 0
+	case 1:
+		distinctRange = float64(numRange)
+	default:
+		distinctRange = rng.Float64() * float64(numRange)
+	}
+	return numRange, distinctRange
 }
 
 // foreignKeyMutator is a MultiStatementMutation implementation which adds

